datafusion/core/src/physical_optimizer/global_requirements.rs

```diff
@@ -0,0 +1,268 @@
+        vec![self.dist_requirement.clone()]
+    }
+
+    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
+        self.input.output_ordering()
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![self.input.clone()]
+    }
+
+    fn required_input_ordering(&self) -> Vec<Option<Vec<PhysicalSortRequirement>>> {
+        vec![self.order_requirement.clone()]
+    }
+
+    fn unbounded_output(&self, children: &[bool]) -> Result<bool> {
+        // Has a single child
+        Ok(children[0])
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        mut children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        Ok(Arc::new(Self::new(
+            children.remove(0), // has a single child
+            self.order_requirement.clone(),
+            self.dist_requirement.clone(),
+        )))
+    }
+
+    fn execute(
+        &self,
+        _partition: usize,
+        _context: Arc<crate::execution::context::TaskContext>,
+    ) -> Result<crate::physical_plan::SendableRecordBatchStream> {
+        unreachable!();
+    }
+
+    fn statistics(&self) -> Statistics {
+        self.input.statistics()
+    }
+}
+
+impl PhysicalOptimizerRule for GlobalRequirements {
+    fn optimize(
+        &self,
+        plan: Arc<dyn ExecutionPlan>,
+        _config: &ConfigOptions,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        match self.mode {
+            RuleMode::Add => require_top_ordering(plan),
+            RuleMode::Remove => plan.transform_up(&|plan| {
+                if let Some(sort_req) =
+                    plan.as_any().downcast_ref::<GlobalRequirementExec>()
+                {
+                    Ok(Transformed::Yes(sort_req.input()))
+                } else {
+                    Ok(Transformed::No(plan))
+                }
+            }),
+        }
+    }
+
+    fn name(&self) -> &str {
+        "GlobalRequirements"
+    }
+
+    fn schema_check(&self) -> bool {
+        true
+    }
+}
+
+/// This functions adds ancillary `GlobalRequirementExec` to the the physical plan, so that
+/// global requirements are not lost during optimization.
+fn require_top_ordering(plan: Arc<dyn ExecutionPlan>) -> Result<Arc<dyn ExecutionPlan>> {
+    let (new_plan, is_changed) = require_top_ordering_helper(plan)?;
+    if is_changed {
+        Ok(new_plan)
+    } else {
+        // Add `GlobalRequirementExec` to the top, with no specified ordering and distribution requirement.
+        Ok(Arc::new(GlobalRequirementExec::new(
+            new_plan,
+            // there is no ordering requirement
+            None,
+            Distribution::UnspecifiedDistribution,
+        )) as _)
+    }
+}
+
+/// Helper function that adds an ancillary `GlobalRequirementExec` to the given plan.
+/// First entry in the tuple is resulting plan, second entry indicates whether any
+/// `GlobalRequirementExec` is added to the plan.
+fn require_top_ordering_helper(
```

Comment:
This is an option also. I think we can add this during initial plan, then in a single pass remove this operator at the end of the optimizer.

Reply:
I thought about this design. I think there are a couple of reason why current design is better

- Currently `GlobalRequirementExec` is not necessarily the top executor in the physical plan. It is the executor above that defines output ordering.
Such as plan below
```sql
ProjectionExec
  OutputRequirementExec
    SortExec (specifies output ordering)
```
The reason for this strategy is that putting `OutputRequirementExec` at the top has couple of problems.
Consider alternative strategy where plan above turns the following
```sql
OutputRequirementExec
  ProjectionExec
    SortExec (specifies output ordering)
```
In this new version `OutputRequirementExec` may not have the necessary column to satisfy ordering desired at the output. Ordering should be satisfied before projection(These column may get lost during projection).
Consider another plan
```sql
  WindowExec(OVER() sum())
    Source (have an ordering)
```
If we were to require output ordering for it above plan would turn to
```
OutputRequirementExec
  WindowExec(OVER() sum())
    Source (have an ordering)
```
However, output ordering in the plan above is accidental. It is not a requirement by the query. And planner should be free to mess this output ordering, if if is helpful.
In short output ordering requirement only when query  contains `ORDER BY` clause at the end. 
In this light, another strategy might be during plan creation we can insert `OutputRequirementExec` when we see `LogicalPlan::Sort`. 

However, in following kind of queries 
```sql
SELECT a, b
FROM (
  SELECT a, b
  FROM table1
  ORDER BY a
)
ORDER BY b
```
This would introduce unnecessary requirement for the subquery.
Hence to require only appropriate absolutely necessary output ordering, we need to traverse physical plan from top to bottom, as long as ordering is maintained. Then put the `OutputRequirementExec` on top of it (this is the current approach). For this we need to have a top down pass on the physical plan. This can be done in `create_initial_plan` stage also. After initial physical plan is created. However, current implementation generally insert corresponding operator(s) for the `logicalPlan` node. Hence it is not easy to integrate top-down traversal code here.

However, I think we can move OutputRequirementExec insertion code on top of here. 
https://github.com/apache/arrow-datafusion/blob/7b12666ec87ea741c3f5b56ddf1647f6d794f9e3/datafusion/core/src/physical_planner.rs#L450

After initial plan is created. What do you think, about this place?

Reply:
Thank you for the writeup and explanation @mustafasrepo -- I agree with your examples. Maybe the core problem is trying to figure out what the "required ordering" from the `ExecutionPlan` where it could be simply an accident of the output of the plan, as in your example

```
  WindowExec(OVER() sum())
    Source (have an ordering)
```

I think the required output order can probably be calculated from  the logical plan, so maybe we should move it there.  🤔 

> However, I think we can move OutputRequirementExec insertion code on top of here.

That makes sense to me. Shall I try it?

Reply:
> I think the required output order can probably be calculated from the logical plan, so maybe we should move it there. 🤔

If we can reliably get the requirements at that stage, this should work too. Feel free to try and we can see if it helps simplify the code